File: Analisis/simulacion/TTT_sim.py
import sys
sys.path.append("../../paquetes/trueskill/")
import pandas as pd
import src as ttt
from importlib import reload

reload(ttt)

from collections import defaultdict
df = pd.read_csv('./simulacion.csv')
df = df[:2500]
#prior_dict = defaultdict(lambda:env.Rating(0,25/3,0,1/100))
#for h_key in set([(h,s) for h, s in zip(df.handicap, df.width) ]):
#    prior_dict[h_key]
#dict(prior_dict)

results = list(df.black_win.map(lambda x: [0,1] if x else [1,0]))
composition = [[[str(w)], [str(b)]] for w, b in zip(df.white, df.black)]
batch = range(0, len(results))
h = ttt.History(composition, results, batch)
#history.trueSkill()
h.convergence()
w_mean = [print(w) for t,w,b in zip(h.times,df.white,df.black) ]

w_mean = [ h.batches[t].posterior(w).mu for t,w,b in zip(h.times,df.white,df.black) ]
b_mean = [ h.batches[t].posterior(b).mu for t,w,b in zip(h.times,df.white,df.black) ]
w_std = [ h.batches[t].posterior(w).sigma for t,w,b in zip(h.times,df.white,df.black) ]
b_std = [ h.batches[t].posterior(b).sigma for t,w,b in zip(h.times,df.white,df.black) ]
evidence = [  h.batches[t].evidences[0] for t in h.times]

res = df.copy()
res["w_mean"] = w_mean
res["w_std"] = w_std
res["b_mean"] = b_mean
res["b_std"] = b_std
res["evidence"] = evidence


res.to_csv("TTT_python.csv", index=False)

File: Analisis/simulacion/sigma_aumenta_jugadores_conocidos/TTT_sim.jl
include("../../../paquetes/trueskill.jl/src/TrueSkill.jl")

using .TrueSkill
global const ttt = TrueSkill
using Test
using CSV
using JLD2
using Dates
using DataFrames
data = DataFrame!(CSV.File("./simulacion.csv"))
#data = data[1:300000,:]

#prior_dict = Dict{String,ttt.Rating}()
#for h_key in Set([(row.handicap, row.width) for row in eachrow(data) ])
#    prior_dict[string(h_key)] = ttt.Rating(0.,25.0/3.,0.,1.0/100)
#end
results = [row.black_win == 0 ? [1,0] : [0, 1] for row in eachrow(data) ]
composition = [[[string(r.white)],[string(r.black)]]  for r in eachrow(data) ]
times = Vector{Int64}()
#times = [row.date for row in eachrow(data)]

#println("Starting TrueSkill")
h = ttt.History(composition, results, times )
w_mean = [ ttt.posterior(h.batches[r], string(data[r,"white"])).mu for r in 1:size(data)[1]]
b_mean = [ ttt.posterior(h.batches[r], string(data[r,"black"])).mu  for r in 1:size(data)[1]]
w_std = [ ttt.posterior(h.batches[r], string(data[r,"white"])).sigma for r in 1:size(data)[1]]
b_std = [ ttt.posterior(h.batches[r], string(data[r,"black"])).sigma for r in 1:size(data)[1]]
evidence = [ h.batches[r].evidences[1] for r in 1:size(data)[1]]

println("Saving Data")
df = DataFrame(white = data[!,:"white"]
              ,black = data[!,:"black"]
              ,w_mean = w_mean
              ,b_mean = b_mean
              ,w_std = w_std
              ,b_std = b_std
              ,evidence = evidence)

#CSV.write("TTT_datos.csv", df; header=true)
CSV.write("Trueskill_datos_sin_handicap.csv", df; header=true)

println("Starting TTT")
ttt.convergence(h)

println("Purifing Data")
w_mean = [ ttt.posterior(h.batches[r], string(data[r,"white"])).mu for r in 1:size(data)[1]]
b_mean = [ ttt.posterior(h.batches[r], string(data[r,"black"])).mu  for r in 1:size(data)[1]]
w_std = [ ttt.posterior(h.batches[r], string(data[r,"white"])).sigma for r in 1:size(data)[1]]
b_std = [ ttt.posterior(h.batches[r], string(data[r,"black"])).sigma for r in 1:size(data)[1]]
evidence = [ h.batches[r].evidences[1] for r in 1:size(data)[1]]

println("Saving Data")
df = DataFrame(white = data[!,"white"]
              ,black = data[!,"black"]
              ,white_skill = data[!,"white_skill"]
              ,black_skill = data[!,"black_skill"]
              ,w_mean = w_mean
              ,b_mean = b_mean
              ,w_std = w_std
              ,b_std = b_std
              ,evidence = evidence)

#CSV.write("TTT_datos.csv", df; header=true)
CSV.write("TTT_datos_sin_handicap.csv", df; header=true)
